packages/vm/src/imports.rs

```diff
@@ -95,156 +82,169 @@ mod errors {
     /// db_scan errors (-2_000_0xx)
     #[cfg(feature = "iterator")]
     pub mod scan {
-        /// An unknown error in the db_scan implementation
-        pub static UNKNOWN: i32 = -2_000_001;
         /// Invalid Order enum value passed into scan
         pub static INVALID_ORDER: i32 = -2_000_002;
     }
 
     /// db_next errors (-2_000_1xx)
     #[cfg(feature = "iterator")]
     pub mod next {
-        /// An unknown error in the db_next implementation
-        pub static UNKNOWN: i32 = -2_000_101;
         /// Iterator with the given ID is not registered
         pub static ITERATOR_DOES_NOT_EXIST: i32 = -2_000_102;
     }
 }
 
-/// Reads a storage entry from the VM's storage into Wasm memory
-pub fn do_read<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32, value_ptr: u32) -> i32 {
-    let key = match read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY) {
-        Ok(data) => data,
-        Err(VmError::RegionLengthTooBig { .. }) => return errors::REGION_READ_LENGTH_TOO_BIG,
-        Err(_) => return errors::REGION_READ_UNKNOWN,
+/// This macro wraps the read_region function for the purposes of the functions below which need to report errors
+/// in its operation to the caller in the WASM runtime.
+/// On success, the read data is returned from the expression.
+/// On failure, an error number is wrapped in `Ok` and returned from the function.
+macro_rules! read_region {
+    ($ctx: expr, $ptr: expr, $length: expr) => {
+        match read_region($ctx, $ptr, $length) {
+            Ok(data) => data,
+            Err(err) => {
+                return Ok(match err {
+                    VmError::RegionLengthTooBig { .. } => errors::REGION_READ_LENGTH_TOO_BIG,
+                    _ => errors::REGION_READ_UNKNOWN,
+                })
+            }
+        }
     };
-    let value: Option<Vec<u8>> = match with_storage_from_context::<S, Q, _, _>(ctx, |store| {
-        store
-            .get(&key)
-            .or_else(|_| Err(make_backend_err("Error reading from backend")))
-    }) {
-        Ok(v) => v,
-        Err(VmError::UninitializedContextData { .. }) => return errors::NO_CONTEXT_DATA,
-        Err(_) => return errors::DB_UNKNOWN,
+}
+
+#[cfg(feature = "iterator")]
+/// This macro wraps the maybe_read_region function for the purposes of the functions below which need to report errors
+/// in its operation to the caller in the WASM runtime.
+/// On success, the optionally read data is returned from the expression.
+/// On failure, an error number is wrapped in `Ok` and returned from the function.
+macro_rules! maybe_read_region {
+    ($ctx: expr, $ptr: expr, $length: expr) => {
+        match maybe_read_region($ctx, $ptr, $length) {
+            Ok(data) => data,
+            Err(err) => {
+                return Ok(match err {
+                    VmError::RegionLengthTooBig { .. } => errors::REGION_READ_LENGTH_TOO_BIG,
+                    _ => errors::REGION_READ_UNKNOWN,
```

Comment:
Ah, you need the macro to be able to return from the surrounding function here (`return i32`, not just the `?` for Result).

It does seem cleaner. I am a bit hesitant on using macros for creeping complexity and debug difficulty. However, if they are private to this module and for this case, it seems fine. I try to avoid them in public API for now.

Reply:
Yeah, I have looked at some other codebases (eg. ink) and when you see so many macros, you don't see the control flow of your code at all. 

I would prefer to polish the public API with just generics until it is super solid. In the future, when we have many users and good idioms for working with everything, then maybe we can add some macros for the "happy path", but not to the point that you need them to use the API.

This was some advise I got early on to avoid powerful Rust magic until you really need it, as you can usually solve it by simplifiying lower down the stack. Not against your use of macros here, just in general.

Reply:
Yeah no worries, i agree with your approach for the most part :)
I don't think you'll even need macros in the public API at all, at least in cosmwasm-vm.

Reply:
Resolving this as I see agreement on the current code in this PR